### DISCUSSION TITLE
Shared directory

### DISCUSSION THREAD

**Discussion by Tommuh** [webpack/webpack#16273] [Q&A] [answered]
Hi guys,

I have searched the webs but I couldn't find a solution for my problem or I simply didn't understand it.
Hopefully you could push me to the right direction.

I have directory for shared components. This directory cannot be placed inside the project because ... reasons.

I tried setting up resolve/alias to the shared components directory and changed all the imports. 
When running "npm start" it find those shared components but it is giving me error messages that it can't find the "node_modules" inside my project-x directory.

Part of Webpack config:

```
      ................
        resolve: {
            modules: [path.resolve(__dirname, 'src'), 'node_modules'],
            alias: {
                library: path.resolve(__dirname, '../components-library/src')
            }
        }, 
        module: {
            rules: [{
                test: /\.(js|jsx)$/,
                exclude: /node_modules/,
                use: {
                    loader: 'babel-loader',
                    options: {
                        configFile: path.resolve(__dirname, '.babelrc')
                    },                    
                }  
            }
        ................
```

Import:

```
import Button from 'library/components/button';
```

My directory structure looks like:

- components-library
   - src
        - components
            - button.js
            - ...js
            - ...js

-  project-x
   - node_modules
   - src
       - index.js
   - package.json
   - webpack.config.js

Error messages are like:

```
ERROR in ../components-library/src/components/button.js 13:0-48
Module not found: Error: Can't resolve 'react/jsx-runtime' in 'D:\.......\components-library\src\components'
```

I'm getting those errors for 'react', 'react-i18next' and many other libraries I use.
What m I doing wrong? Could you please help met out.

I have tried a few things but not worth mentioning.

Thanks!

**Comment by alexander-akait** [comment]
Do you have installed `react`? If you want to use `node_modules` directory from `project-x`, you need setup https://webpack.js.org/configuration/resolve/#resolvemodules

**Comment by Tommuh** [comment]
Yes I do have react installed and have resolve modules configured. I forgot to mention I'm using Webpack 5.
When I drop the contents of "components-library" inside my project everything works as expected. 

This is the full webpack config located in project-x:

```
{
        mode: 'production',
        entry: { 
            index: path.resolve(__dirname, 'src', 'index.js')
        },
        output: {
            filename: 'modules/[name].[contenthash].js',
            path: path.resolve(__dirname, 'build'),
            clean: true
        },
        plugins: [
            new HtmlWebpackPlugin({
                template: path.resolve(__dirname, 'public', 'index.html')
            }),
            new MiniCssExtractPlugin({
                filename: 'css/[name].[contenthash].css',
                chunkFilename: 'css/[id].[contenthash].css'
            }),
            new CopyPlugin({
                patterns: [{ 
                    from: path.resolve(__dirname, 'public/img'),
                    to: 'img' 
                }]
            }),   
            new Dotenv({
                path: path.resolve(__dirname, '.env.' + env.ENVIRONMENT),
                silent: false,
                expand: true,
                defaults: path.resolve(__dirname, '.env.defaults')
            })    
        ],
        resolve: {
            modules: [path.resolve(__dirname, 'src'), 'node_modules'],
            alias: {
                library: path.resolve(__dirname, '../components-library/src')
            }
        },    
        module: {
            rules: [{
                test: /\.(css|s[ac]ss)$/,
                use: [
                    'style-loader', 
                    'css-loader', 
                    {
                        loader: 'sass-loader',
                        options: {
                            sourceMap: true,
                            warnRuleAsWarning: false,
                            sassOptions: {
                                quietDeps: true
                            }
                        }
                    }
                ]
            }, {
                test: /\.(js|jsx)$/,
                exclude: /node_modules/,
                use: ['babel-loader']   
            }, {
                test: /\.(woff|woff2|ttf)$/i,
                type: 'asset/resource',
                generator: {
                    filename: 'fonts/[name][ext][query]'
                }
            }, {
                test: /\.(png|svg|jpg|jpeg|gif)$/i,    
                type: 'asset/resource',
                generator: {
                    filename: 'img/[name][ext][query]'
                }
            }]
        },
        optimization: {
            minimize: true,
            minimizer: [
                new TerserPlugin({
                    terserOptions: {
                        format: {
                            comments: false
                        }
                    },
                    extractComments: false,            
                    parallel: true
                })             
            ],
            mergeDuplicateChunks: false,
            removeEmptyChunks: true,
            moduleIds: 'deterministic',
            splitChunks: { 
                chunks: 'all',
                cacheGroups: {
                    commons: {
                        name: 'library/shared',
                        test: /[\\/](components)[\\/]/,
                        chunks: 'all'
                    },
                    vendor: {
                        name: 'library/vendors',
                        test: /[\\/]node_modules[\\/]/,
                        chunks: 'all',
                        priority: 10,
                        enforce: true
                    }
                }
            },
            runtimeChunk: {
                name: (entrypoint) => `runtimechunk~${entrypoint.name}`
            }
        }
    }
```

**Reply by alexander-akait** [reply]
Put `modules: [path.resolve(__dirname, 'src'), 'node_modules']` here directory where you have another `node_modules`

**Reply by Tommuh** [reply]
Thank you for your quick reply. I will try it out tonight.

Currently I don't have a node_modules inside my "components-library". 
But it makes senses to move the packages to that directory that have something to do with "components-library" and remove those from "project-x".

**Reply by alexander-akait** [reply]
Yes, you can move modules directory to the root of your project

**Reply by Tommuh** [reply]
Well both components-library and project-x are not standing together in a root directory. Obviously standing side by side, but it is not that I can place node_modules at the same level. Or do you mean something else?

**Reply by alexander-akait** [reply]
Why you can't do it? There is example of a monorepo to have many projects in one repo https://github.com/Quramy/npm-ts-workspaces-example

**Reply by Tommuh** [reply]
Well I want to have less files for each repo. My current project has 1000+ files (node_modules excluded). This project is 1/4 done. 
I'm starting the next project parallel to my current project. This new project will probably have the same amount of files maybe more.

That is why I started splitting my source ...

**Reply by Tommuh** [reply]
Besides the monorepo way - is there an other solution to solve this?

**Reply by alexander-akait** [reply]
I think better to ask this on stackoverflow, we are bundled, sorry

I simply cannot give a sufficiently detailed and broad answer in all directions.

**Reply by Tommuh** [reply]
Hmm no problem - thanks for trying to help me.

**Comment by Tommuh** [comment, marked as answer]
Eventually solved it with git submodules.

I have a main repo called "workspaces".
This "workspaces" holds my node_modules, package.json and my webpack files and a directory for the submodules called "workspaces".

My projects all need the same packages. 
If one of those projects have a unique package in can define them separately.

![image](https://user-images.githubusercontent.com/2089879/191375297-c67da14e-46ca-4150-a3e1-bccefb13dbe1.png)

Inside workspaces:

![image](https://user-images.githubusercontent.com/2089879/191374610-f2914d7f-54a6-41c5-a96a-d234582f4012.png)

Inside on of those workspace:

![image](https://user-images.githubusercontent.com/2089879/191374764-559afb7a-1267-4756-99f6-ac09b27fa316.png)

If someone is encountering the same problem and is interested in configs you can send me a message.
